executor/metric_reader.go

```diff
@@ -52,29 +55,55 @@ func (e *MetricRetriever) retrieve(ctx context.Context, sctx sessionctx.Context)
 		return nil, err
 	}
 	e.tblDef = tblDef
-	// TODO: Get query range from plan instead of use default range.
-	queryRange := e.getDefaultQueryRange(sctx)
-	queryValue, err := e.queryMetric(ctx, sctx, queryRange)
+	queryRange := e.getQueryRange(sctx)
+	quantiles, err := e.extractor.GetQuantiles()
 	if err != nil {
 		return nil, err
 	}
+	rows := make([][]types.Datum, 0)
+	for _, quantile := range quantiles {
+		if quantile == 0 {
+			quantile = tblDef.Quantile
+		}
+		queryValue, err := e.queryMetric(ctx, sctx, queryRange, quantile)
+		if err != nil {
+			return nil, err
+		}
 
-	fullRows = e.genRows(queryValue, queryRange)
-	if len(e.outputCols) == len(e.table.Columns) {
-		return
-	}
-	rows := make([][]types.Datum, len(fullRows))
-	for i, fullRow := range fullRows {
-		row := make([]types.Datum, len(e.outputCols))
-		for j, col := range e.outputCols {
-			row[j] = fullRow[col.Offset]
+		fullRows = e.genRows(queryValue, queryRange, quantile)
+		if len(e.outputCols) == len(e.table.Columns) {
+			rows = append(rows, fullRows...)
+			continue
+		}
+		for _, fullRow := range fullRows {
+			row := make([]types.Datum, len(e.outputCols))
+			for j, col := range e.outputCols {
+				row[j] = fullRow[col.Offset]
+			}
+			rows = append(rows, row)
 		}
-		rows[i] = row
 	}
 	return rows, nil
 }
 
-func (e *MetricRetriever) queryMetric(ctx context.Context, sctx sessionctx.Context, queryRange promv1.Range) (pmodel.Value, error) {
+func (e *MetricRetriever) queryMetric(ctx context.Context, sctx sessionctx.Context, queryRange promv1.Range, quantile float64) (pmodel.Value, error) {
+	failpoint.Inject("mockMetricRetrieverQueryPromQL", func() {
```

Comment:
Prefer to construct test data in `_test` and pass it here via `context.Context`.
like:
```
failpoint.InjectContext(ctx, "mockMetricRetrieverQueryPromQL", func() {
    return ctx.Value("__mockMetricsData").(pmodel.SamplePair)
}
```
Put the mock data and test case togother will help review easily.

Reply:
Great, done.